drivers/serial/uart_samd.c

```diff
@@ -0,0 +1,287 @@
+			       const int size)
+{
+	SercomUsart *const regs = DEV_CFG(dev)->regs;
+
+	if (regs->INTFLAG.bit.RXC) {
+		u8_t ch = regs->DATA.reg;
+
+		if (size >= 1) {
+			*rx_data = ch;
+			return 1;
+		} else {
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+static int uart_samd_irq_is_pending(struct device *dev)
+{
+	SercomUsart *const regs = DEV_CFG(dev)->regs;
+
+	return (regs->INTENSET.reg & regs->INTFLAG.reg) != 0;
+}
+
+static int uart_samd_irq_update(struct device *dev) { return 1; }
+
+static void uart_samd_irq_callback_set(struct device *dev,
+				       uart_irq_callback_t cb)
+{
+	struct uart_samd_dev_data *const dev_data = DEV_DATA(dev);
+
+	dev_data->cb = cb;
+}
+#endif
+
+static const struct uart_driver_api uart_samd_driver_api = {
+	.poll_in = uart_samd_poll_in,
+	.poll_out = uart_samd_poll_out,
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_read = uart_samd_fifo_read,
+	.irq_tx_disable = uart_samd_irq_tx_disable,
+	.irq_rx_enable = uart_samd_irq_rx_enable,
+	.irq_rx_disable = uart_samd_irq_rx_disable,
+	.irq_rx_ready = uart_samd_irq_rx_ready,
+	.irq_is_pending = uart_samd_irq_is_pending,
+	.irq_update = uart_samd_irq_update,
+	.irq_callback_set = uart_samd_irq_callback_set,
+#endif
+};
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+#define UART_SAMD_INIT_DEFN(addr, n)					       \
+	static int uart_samd_init_##n(struct device *dev)		       \
+	{								       \
+		uart_samd_init(dev);					       \
+		IRQ_CONNECT(ATMEL_SAMD_USART_##addr##_IRQ_0,		       \
+			    ATMEL_SAMD_USART_##addr##_IRQ_0_PRIORITY,	       \
+			    uart_samd_isr, DEVICE_GET(uart_samd_##n), 0);      \
+		irq_enable(ATMEL_SAMD_USART_##addr##_IRQ_0);		       \
+		return 0;						       \
+	}
+#else
+#define UART_SAMD_INIT_DEFN(addr, n)					       \
+	static int uart_samd_init_##n(struct device *dev)		       \
+	{								       \
+		return uart_samd_init(dev);				       \
+	}
+#endif
+
+#define UART_SAMD_DEVICE_INIT(addr, n)					       \
+	static int uart_samd_init_##n(struct device *dev);		       \
+									       \
+	static struct uart_samd_dev_data uart_samd_data_##n = {};	       \
+									       \
+	DEVICE_AND_API_INIT(uart_samd_##n, ATMEL_SAMD_USART_##addr##_LABEL,    \
+			    uart_samd_init_##n, &uart_samd_data_##n,	       \
+			    &uart_samd_config_##n, PRE_KERNEL_1,	       \
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		       \
+			    &uart_samd_driver_api);			       \
+	UART_SAMD_INIT_DEFN(addr, n)
+
+/* SERCOM0 */
+#ifdef ATMEL_SAMD_USART_42000800_BASE_ADDRESS
+
+static const struct uart_samd_dev_cfg uart_samd_config_0 = {
+	.regs = (SercomUsart *)ATMEL_SAMD_USART_42000800_BASE_ADDRESS,
+	.pin_rx = {0, 11, PORT_PMUX_PMUXE_C_Val},
+	.pin_tx = {0, 10, PORT_PMUX_PMUXE_C_Val},
+	.baudrate = ATMEL_SAMD_USART_42000800_CURRENT_SPEED,
+	.pm_apbcmask = PM_APBCMASK_SERCOM0,
+	.gclk_clkctrl_id = GCLK_CLKCTRL_ID_SERCOM0_CORE,
+	.ctrla = SERCOM_USART_CTRLA_RXPO(3) | SERCOM_USART_CTRLA_TXPO(1),
+};
+
+UART_SAMD_DEVICE_INIT(42000800, 0)
```

Comment:
we shouldn't be hard coding 42000800

Reply:
Hi @galak what would you recommend?  As background, the device tree has lines like:

sercom0: uart@42000800 {

which generates constants like ATMEL_SAMD_USART_42000800_BASE_ADDRESS and ATMEL_SAMD_USART_42000800_LABEL.  I use the constant above and the preprocessor to generate these names inside the device initializer.

The closest I can find is the stm32 driver which uses a fixup file like:

#define CONFIG_UART_STM32_PORT_1_NAME           ST_STM32_USART_40013800_LABEL

but it seems a pity to use a fixup file to convert device tree names to human names.